nano/secure/stream_filter.cpp

```diff
@@ -0,0 +1,63 @@
+#include <nano/crypto_lib/random_pool.hpp>
+#include <nano/lib/locks.hpp>
+#include <nano/secure/stream_filter.hpp>
+
+nano::stream_filter::stream_filter (size_t capacity_a) :
+capacity (capacity_a),
+items (capacity_a, item_key_t{ 0 })
+{
+	nano::random_pool::generate_block (key, key.size ());
+}
+
+bool nano::stream_filter::operator() (bool & error_a, nano::stream & stream_a)
+{
+	bool existed;
+	auto digest (hash (error_a, stream_a));
+	if (!error_a)
+	{
+		nano::lock_guard<std::mutex> lock (mutex);
+		assert (capacity > 0 && items.size () == capacity);
+		size_t index (digest % items.size ());
+		auto & element (items[index]);
+		existed = element == digest;
+		if (!existed)
+		{
+			// Replace likely old element with a new one
+			element = digest;
+		}
+	}
+	return existed;
+}
+
+nano::stream_filter::item_key_t nano::stream_filter::hash (bool & error_a, nano::stream & stream_a) const
+{
+	CryptoPP::SipHash<2, 4, true> siphash (key, key.size ());
+	nano::uint128_union digest;
+	try
+	{
+		std::array<uint8_t, 8> buffer;
+		size_t bytes_read;
+		size_t total_read{ 0 };
+		do
+		{
+			bytes_read = stream_a.sgetn (buffer.data (), buffer.size ());
+			total_read += bytes_read;
+			siphash.Update (buffer.data (), bytes_read);
+		} while (bytes_read == buffer.size ());
+
+		// Rewind stream to previous state
+		stream_a.pubseekoff (-total_read, std::ios_base::cur);
+		siphash.Final (digest.bytes.data ());
```

Comment:
Since call sites have the raw buffer, I wonder if passing `int8_t const * buffer_a, size_t size_a` instead of the stream would allow a single `siphash.Update` call for the entire buffer, instead of looping and rewinding at the end.

Reply:
Had this initially, there is even a CalculateDigest that does an "all at once" update. `message_parser::deserialize_buffer` first deserializes the header and there doesn't seem to be any `header::size ()` to subtract from the total size, should I make it?

Reply:
Right, and skip accordingly, e.g `buffer_a + header::size()`

Reply:
Done, also now padding the buffer to ensure it has a size multiple of 8 bytes.

Reply:
Reverted in 9d13191